### PR TITLE
feat(model): Add new model to store settings

### PR DESCRIPTION
**Pull request by kmala**

### CODE REVIEW

File: rootfs/api/migrations/0012_auto_20160810_0315.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
```
Comment (helgi) [inline rootfs/api/migrations/0012_auto_20160810_0315.py]: Should probably try to split this into 2 migrations, the stuff that's coming from all those other models and then yours.

Maybe make a separate PR for that, or i can

**Comment by helgi** [comment]
Why is this a whole new model instead of under the Config model?

File: rootfs/api/migrations/0012_auto_20160810_0315.py (hunk 1 above)
Comment (helgi) [inline rootfs/api/migrations/0012_auto_20160810_0315.py]: https://github.com/deis/controller/pull/967

**Comment by kmala** [comment]
Currently App maps to release and release object is aggregation of config object and build object. So any change to the config or the build would trigger a new release and hence a new deploy of the app and we store all the configuration values in the config object.
But sometimes we may not need/want to create a release for few configurations changes like (enable or disable of maintenance mode or setting value for the deployments strategy or setting KUBERNETES_DEPLOYMENTS_REVISION_HISTORY_LIMIT(https://github.com/deis/controller/blob/master/rootfs/api/models/app.py#L473))
So i had created a new model appsettings which maps to App model and stores different configuration data which doesn't need to do a release.

File: rootfs/api/models/__init__.py [hunk 2]
```diff
@@ -142,6 +143,12 @@ def _log_config_updated(**kwargs):
     config.app.log("config {} updated".format(config))
 
 
+def _log_app_settings_updated(**kwargs):
+    appSettings = kwargs['instance']
+    # log only to the controller; this event will be logged in the release summary
+    appSettings.app.log("application setings {} updated".format(appSettings))
```
Comment (bacongobbler) [inline rootfs/api/models/__init__.py]: typo: settings

File: rootfs/api/models/app.py [hunk 3]
```diff
@@ -822,3 +833,20 @@ def _build_env_vars(self, build_type, version, image, envs):
         # merge envs on top of default to make envs win
         default_env.update(envs)
         return default_env
+
+    def set_maintenance(self, maintenance):
```
Comment (helgi) [inline rootfs/api/models/app.py]: Debating if this should live here or in the AppSettings model. What's the benefit of keeping it in `App`?

File: rootfs/scheduler/__init__.py [hunk 4]
```diff
@@ -259,13 +261,14 @@ def _get_deploy_batches(self, steps, desired):
...
 
-    def _update_application_service(self, namespace, name, app_type, port, routable=False):
+    def _update_application_service(self, namespace, name, app_type, port, routable=False, maintenance=False):  # noqa
         """Update application service with all the various required information"""
         service = self.get_service(namespace, namespace).json()
         old_service = service.copy()  # in case anything fails for rollback
 
         try:
             # Update service information
+            service['metadata']['annotations']['router.deis.io/maintenance'] = str(maintenance)
```
Comment (helgi) [inline rootfs/scheduler/__init__.py]: You already are doing this with `set_maintenance` what's the benefit here?

Comment (kmala) [inline rootfs/scheduler/__init__.py]: if the user sets the appsettings first and then does a deploy set_maintenance doesn't get called..set_maintenance is called only during the save of the record and if the value is changed.

File: rootfs/api/models/app.py (hunk 3 above)
Comment (kmala) [inline rootfs/api/models/app.py]: not much..this can be moved.

Comment (helgi) [inline rootfs/api/models/app.py]: Lets just move it, keep the models focused and tidy

File: rootfs/scheduler/__init__.py (hunk 4 above)
Comment (helgi) [inline rootfs/scheduler/__init__.py]: Currently the `Service` is always created when the application is, so the resource is always around, that should mean the `Service` is ready to go before a deploy. That's how `Domains` work, so I'm just making sure things are all in sync here.

**Comment by helgi** [comment]
Don't forget to add this model into the `load_db_to_k8s.py`

**Comment by kmala** [comment]
doesn't `load_db_to_k8s.py` of the app take care of the setting? am i missing something?

**Comment by helgi** [comment]
Look at https://github.com/deis/controller/blob/master/rootfs/api/management/commands/load_db_state_to_k8s.py#L50

File: rootfs/api/models/appsettings.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            if new_maintenance is None and prev_maintenance is not None:
+                setattr(self, 'maintenance', prev_maintenance)
+            elif prev_maintenance != new_maintenance:
+                self.set_maintenance(new_maintenance)
+                summary += "{} changed maintenance mode from {} to {}".format(self.owner, prev_maintenance, new_maintenance)  # noqa
+        except Exception as e:
+            self.delete()
+            raise DeisException(str(e)) from e
+
+        if not summary and previous_settings:
```
Comment (bacongobbler) [inline rootfs/api/models/appsettings.py]: why are we checking for a summary, but not doing anything with it? Shouldn't we just check if the current settings is != old settings?

Comment (kmala) [inline rootfs/api/models/appsettings.py]: i wanna log the summary but i think i missed that.

File: rootfs/api/tests/test_maintenance.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import requests_mock
```
Comment (bacongobbler) [inline rootfs/api/tests/test_maintenance.py]: can you rename this file to `test_app_settings.py` since that is the object we're testing here?

File: rootfs/api/tests/test_maintenance.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+from django.core.cache import cache
+from django.contrib.auth.models import User
+from rest_framework.authtoken.models import Token
+
+from api.models import App
+from api.tests import adapter, DeisTransactionTestCase
+
+
+@requests_mock.Mocker(real_http=True, adapter=adapter)
+class TestMaintenance(DeisTransactionTestCase):
```
Comment (bacongobbler) [inline rootfs/api/tests/test_maintenance.py]: same here, can we change this to `TestAppSettings`

**Comment by bacongobbler** [comment]
I see all the tests for maintenance mode, but could we please add unit tests for all the intricate edge cases in the AppSettings.save() function? For example, when the summary is empty and previous_settings is not none, we delete the current settings and raise AlreadyExists?

File: rootfs/scheduler/__init__.py [hunk 8]
```diff
@@ -176,13 +177,14 @@ def _get_deploy_batches(self, steps, desired):
 
         return batches
 
-    def _update_application_service(self, namespace, name, app_type, port, routable=False):
+    def _update_application_service(self, namespace, name, app_type, port, routable=False, maintenance=False):  # noqa
```
Comment (bacongobbler) [inline rootfs/scheduler/__init__.py]: perhaps we should rewrite this to read for an `annotations` kwarg and attach that to the service metadata instead of making this an N-length method, where N is the number of supported annotations in the router.

File: rootfs/scheduler/__init__.py [hunk 9]
```diff
@@ -89,6 +89,7 @@ def deploy(self, namespace, name, image, entrypoint, command, **kwargs):  # noqa
 
         app_type = kwargs.get('app_type')
         routable = kwargs.get('routable', False)
```
Comment (bacongobbler) [inline rootfs/scheduler/__init__.py]: same thought as below IRT an `annotations` kwarg.

File: rootfs/api/views.py [hunk 10]
```diff
@@ -296,6 +296,11 @@ def restart(self, *args, **kwargs):
         return Response(pagination, status=status.HTTP_200_OK)
 
 
+class AppSettingsViewSet(AppResourceViewSet):
```
Comment (bacongobbler) [inline rootfs/api/views.py]: we need to implement a `get_object` method or the only way to retrieve an AppSetting will be through the UUID. See the CertificateViewSet's `get_object` method for an example.

File: rootfs/scheduler/__init__.py (hunk 8 above)
Comment (kmala) [inline rootfs/scheduler/__init__.py]: currently there is only one annotation and one label which is updated through controller.

Comment (helgi) [inline rootfs/scheduler/__init__.py]: Could use https://github.com/deis/controller/blob/master/rootfs/api/models/__init__.py#L52-L94 if we want to go as far as naming it after a particular component - That's what `Domains` currently uses

**Comment by kmala** [comment]
The test are already present to test for AlreadyExists and  previous_settings is not None

File: rootfs/api/models/app.py [hunk 11]
```diff
@@ -389,6 +394,7 @@ def scale(self, user, structure):  # noqa
 
     def _scale_pods(self, scale_types):
         release = self.release_set.latest()
+        appSettings = self.appsettings_set.latest()
```
Comment (bacongobbler) [inline rootfs/api/models/app.py]: this use_underscores for the variable name rather than camelCase.

File: rootfs/api/models/app.py [hunk 12]
```diff
@@ -23,6 +23,7 @@
 from api.models.release import Release
 from api.models.config import Config
 from api.models.domain import Domain
+from api.models.appsettings import AppSettings
```
Comment (bacongobbler) [inline rootfs/api/models/app.py]: I almost feel as if this should be part of another package dir...

```
from api.models.app.settings import Settings
```

Any thoughts on the reasoning behind this package naming scheme? I mean it's directly tied to the app model...

Comment (bacongobbler) [inline rootfs/api/models/app.py]: this is getting into bikeshed territory so disregard.

File: rootfs/scheduler/__init__.py (hunk 9 above)
Comment (kmala) [inline rootfs/scheduler/__init__.py]: `routable` is added as label not as annotation

**Comment by bacongobbler** [comment]
> The test are already present to test for AlreadyExists and previous_settings is not None

I'm sorry but I don't see any `assertRaises` calls for either `AlreadyExists` or `DeisException` in `test_maintenance.py`, nor any tests that the setting was deleted when said if block was encountered. Where are these unit tests?

File: rootfs/api/models/appsettings.py (hunk 5 above)
Comment (bacongobbler) [inline rootfs/api/models/appsettings.py]: m0ar tests

File: rootfs/scheduler/__init__.py (hunk 9 above)
Comment (bacongobbler) [inline rootfs/scheduler/__init__.py]: correct, but I'm referring to other annotations such as IP whitelisting in https://github.com/deis/controller/issues/240.

**Comment by kmala** [comment]
I am checking for the response codes AlreadyExists(409)

**Comment by bacongobbler** [comment]
Cool, I missed that. Sorry. I was looking for the error itself.

File: rootfs/api/models/appsettings.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            self.delete()
+            raise AlreadyExists("{} changed nothing".format(self.owner))
+        self.app.log('summary of app setting changes: {}'.format(summary), logging.DEBUG)
```
Comment (bacongobbler) [inline rootfs/api/models/appsettings.py]: nitpick but I think we should just log the summary in the app logs, similar to when we [dump the release summary](https://github.com/deis/controller/blob/master/rootfs/api/models/__init__.py#L135-L136).

File: rootfs/api/models/__init__.py [hunk 14]
```diff
@@ -142,6 +143,12 @@ def _log_config_updated(**kwargs):
     config.app.log("config {} updated".format(config))
 
 
+def _log_app_settings_updated(**kwargs):
+    appSettings = kwargs['instance']
+    # log only to the controller; this event will be logged in the release summary
+    appSettings.app.log("application settings {} updated".format(appSettings))
```
Comment (bacongobbler) [inline rootfs/api/models/__init__.py]: this is logged also to the application, which means the user will see this log as well as the summary. Do you mean to call `logger.info()` instead?

Comment (kmala) [inline rootfs/api/models/__init__.py]: This just says that settings have been updated but not what all has changed. app.log will call logger.log as with level as info
If Debug mode is on then user will see the full details of what all has changed which is what i had done above.

**Comment by bacongobbler** [comment]
if you wanted to move `routable` to appsettings,  now is the time.

**Comment by kmala** [comment]
started making changes, will be changing controller sdk, workflow-cli and controller..that should be it right?

**Comment by bacongobbler** [comment]
yep! That should be the only places required.

File: rootfs/api/migrations/0012_auto_20160810_1603.py [hunk 15]
```diff
@@ -9,7 +9,7 @@
 class Migration(migrations.Migration):
```
Comment (helgi) [inline rootfs/api/migrations/0012_auto_20160810_1603.py]: Since you deleted the other one, may want to rename this to 0011

Comment (kmala) [inline rootfs/api/migrations/0012_auto_20160810_1603.py]: changed

File: rootfs/api/models/app.py (hunk 12 above)
Comment (helgi) [inline rootfs/api/models/app.py]: Agree but we can refactor later

File: rootfs/api/models/app.py [hunk 16]
```diff
@@ -424,6 +430,7 @@ def _scale_pods(self, scale_types):
                 'app_type': scale_type,
                 'build_type': release.build.type,
                 'healthcheck': healthcheck,
+                'annotations': {'maintenance': app_settings.maintenance},
```
Comment (helgi) [inline rootfs/api/models/app.py]: This is rather unclear since this is service only but all resource objects in k8s takes annotations

Comment (kmala) [inline rootfs/api/models/app.py]: do you want it to be renamed to service_annotations?

Comment (helgi) [inline rootfs/api/models/app.py]: Yeah think that'd be better given the scope of that `kwargs`
